### PR TITLE
a system for caching user loads

### PR DESCRIPTION
**Pull request by maxtaco**
- we're not caching users, but rather UserPlusAllKeys objects, which we can do some interesting things with
- some tests

### CODE REVIEW

File: go/libkb/cached_user_loader.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+		}
+		if fresh {
+			return upk.DeepCopy(), nil, nil
+		}
+		if info != nil {
+			info.TimedOut = true
+		}
+	}
+
+	if upk != nil {
```
Comment (mlsteele) [inline go/libkb/cached_user_loader.go]: This has the same if condition `upk != nil` as the above block. Combine 'em?

File: go/libkb/cached_user_loader.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+		var sigHints *SigHints
+		sigHints, err = LoadSigHints(arg.UID, u.G())
+		if err != nil {
+			return nil, nil, err
+		}
+
+		var leaf *MerkleUserLeaf
+		leaf, err = lookupMerkleLeaf(u.G(), arg.UID, true, sigHints)
```
Comment (mlsteele) [inline go/libkb/cached_user_loader.go]: Check this `err` from `lookupMerkleLeaf`

File: go/libkb/upak.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package libkb
+
+// UPAK = "User Plus All Keys"
+
+import (
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+)
+
+// BaseProofSet creates a basic proof set for a user with their
+// keybase and uid proofs and any pgp fingerpring proofs.
```
Comment (mlsteele) [inline go/libkb/upak.go]: typo fingerpring

File: go/protocol/keybase1/extras.go [hunk 4]
```diff
@@ -757,3 +758,33 @@ func (b TLFIdentifyBehavior) WarningInsteadOfErrorOnBrokenTracks() bool {
...
+func (u UserPlusKeys) GetUID() UID {
+	return u.Uid
+}
+
+func (u UserPlusKeys) GetName() string {
+	return u.Username
+}
+
+func (u *UserPlusAllKeys) DeepCopy() *UserPlusAllKeys {
+	handle := &codec.MsgpackHandle{
```
Comment (mlsteele) [inline go/protocol/keybase1/extras.go]: ha, ok.

File: go/libkb/cached_user_loader.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	ret = &tmp
+	ret.Base.Uvv.CachedAt = keybase1.ToTime(u.G().Clock().Now())
+	u.Lock()
+	u.m[arg.UID.String()] = ret
+	u.Unlock()
+
+	return ret, user, nil
+}
+
+func (u *CachedUserLoader) Load(arg LoadUserArg) (ret *keybase1.UserPlusAllKeys, user *User, err error) {
```
Comment (mlsteele) [inline go/libkb/cached_user_loader.go]: Maybe a comment describing the fallback behavior would be nice
`// Load a user from the local cache, falls back to LoadUser`

**Comment by maxtaco** [comment]
Ok, this is all ready, and has #4900 merged in

File: go/libkb/cached_user_loader.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	u.Lock()
+	u.m[arg.UID.String()] = ret
+	u.Unlock()
+
+	return ret, user, nil
+}
+
+// Load a UserPlusAllKeys from the local cache, falls back to LoadUser, and cache the user.
+// Can only perform lookups by UID. Will return a non-nil UserPlusAllKeys, or a non-nil error,
+// but never both non-nill, nor never both nil. If we had to do a full LoadUser as part of the
```
Comment (mlsteele) [inline go/libkb/cached_user_loader.go]: typo `non-nill`, doesn't matter though

**Pull request merged**